### PR TITLE
Convert sql_bank to use cockroach

### PR DESCRIPTION
**Pull request by vivekmenezes**

### CODE REVIEW

**Comment by MBerhault** [comment]
LGTM

**Comment by tamird** [comment]
the 1st and 2nd commits should be squashed - there's some weird churn in there

File: examples/sql_bank/main.go [hunk 1]
```diff
@@ -142,25 +155,31 @@ func verifyBank(db *sql.DB) {
 }
 
 func main() {
-	db, err := sql.Open("postgres", "dbname=postgres sslmode=disable")
+	security.SetReadFileFn(securitytest.Asset)
+	serv := server.StartTestServer(nil)
+	defer serv.Stop()
+
+	url := "https://root@" + serv.ServingAddr() + "?certs=test_certs"
```
Comment (tamird) [inline examples/sql_bank/main.go]: can we preserve the ability to run this on postgres? a command line flag or environment variable would do.

Comment (bdarnell) [inline examples/sql_bank/main.go]: We should also be able to run it against arbitrary cockroach servers (e.g. a multi-node cluster), not just the in-process TestServer.

File: examples/sql_bank/main.go [hunk 2]
```diff
@@ -49,10 +52,15 @@ func moveMoney(db *sql.DB) {
 		if err != nil {
 			log.Fatal(err)
 		}
-		if _, err = tx.Exec("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE"); err != nil {
-			log.Fatal(err)
-		}
+		startTime := time.Now()
+		// TODO(vivek): Query is very slow, needs investigation.
```
Comment (petermattis) [inline examples/sql_bank/main.go]: We don't optimize `IN` queries right now. See #2140. You could rewrite this as:

```
SELECT id, balance from accounts WHERE id = $1 OR id = $2
```

I'm planning on getting the "point" query support done this week.

File: examples/sql_bank/main.go [hunk 3]
```diff
@@ -142,25 +155,31 @@ func verifyBank(db *sql.DB) {
...
+	defer serv.Stop()
+
+	url := "https://root@" + serv.ServingAddr() + "?certs=test_certs"
+	db, err := sql.Open("cockroach", url)
 	if err != nil {
 		log.Fatal(err)
 	}
-
 	if _, err = db.Exec("CREATE DATABASE bank"); err != nil {
 		if pqErr, ok := err.(*pq.Error); !ok || pqErr.Code.Name() != "duplicate_database" {
```
Comment (petermattis) [inline examples/sql_bank/main.go]: This won't be needed for running against cockroach.

File: sql/txn.go [hunk 4]
```diff
@@ -30,6 +30,9 @@ func (p *planner) BeginTransaction(n *parser.BeginTransaction) (planNode, error)
 // CommitTransaction commits a transaction.
 func (p *planner) CommitTransaction(n *parser.CommitTransaction) (planNode, error) {
 	err := p.txn.Commit()
+	if err != nil {
+		_ = p.txn.Rollback()
```
Comment (petermattis) [inline sql/txn.go]: I thought we had agreed that if a commit failed a rollback wouldn't be necessary? @tschottdorf?

File: examples/sql_bank/main.go (hunk 1 above)
Comment (vivekmenezes) [inline examples/sql_bank/main.go]: Thats for another change. I just wanted this to be a "convert to cockroach" change.

yeah I can preserve the postgres behavior if you'd like.

**Comment by vivekmenezes** [comment]
There are no longer any significant performance issues with this PR.

File: examples/sql_bank/main.go [hunk 5]
```diff
@@ -49,35 +82,9 @@ func moveMoney(db *sql.DB) {
 		if err != nil {
 			log.Fatal(err)
 		}
-		if _, err = tx.Exec("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE"); err != nil {
-			log.Fatal(err)
-		}
-		rows, err := tx.Query("SELECT id, balance FROM accounts WHERE id IN ($1, $2)", from, to)
```
Comment (petermattis) [inline examples/sql_bank/main.go]: We should really keep the single query to retrieve both balances. It's slow now, but will be more efficient in a few days than separate queries. It's good to know that doing the separate queries fixes the performance.

Comment (petermattis) [inline examples/sql_bank/main.go]: Another reason for keeping the currently single query, it will make it really obvious when I fix #2140.

Comment (vivekmenezes) [inline examples/sql_bank/main.go]: hey good to know. I'll revert this commit then.

**Comment by tamird** [comment]
@vivekmenezes sql_bank got checked in accidentally looks like

File: examples/sql_bank/main.go [hunk 6]
```diff
@@ -26,14 +26,18 @@ import (
...
 	"github.com/cockroachdb/cockroach/util/log"
 )
 
 const (
 	maxTransfer = 999
 	numAccounts = 999
-	concurrency = 25
+	concurrency = 5
 	aggregate   = true
+	usePostgres = false
```
Comment (petermattis) [inline examples/sql_bank/main.go]: You should make this a flag. The `flags` package is really easy to use. Heck, these should all be flags, but at least the `usePostgres` one to start.

**Comment by petermattis** [comment]
LGTM

File: examples/sql_bank/main.go [hunk 7]
```diff
@@ -142,25 +155,44 @@ func verifyBank(db *sql.DB) {
 }
 
 func main() {
-	db, err := sql.Open("postgres", "dbname=postgres sslmode=disable")
-	if err != nil {
-		log.Fatal(err)
+	security.SetReadFileFn(securitytest.Asset)
+	serv := server.StartTestServer(nil)
```
Comment (tamird) [inline examples/sql_bank/main.go]: this should be under `if !usePostgres`

File: examples/sql_bank/main.go [hunk 8]
```diff
@@ -49,10 +53,13 @@ func moveMoney(db *sql.DB) {
...
 		}
-		if _, err = tx.Exec("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE"); err != nil {
-			log.Fatal(err)
-		}
+		startTime := time.Now()
+		// Query is very slow and will be fixed by https://github.com/cockroachdb/cockroach/issues/2140
 		rows, err := tx.Query("SELECT id, balance FROM accounts WHERE id IN ($1, $2)", from, to)
+		elapsed := time.Now().Sub(startTime)
+		if elapsed > 10*time.Millisecond {
+			log.Info("query took %v", elapsed)
```
Comment (tamird) [inline examples/sql_bank/main.go]: should log the query too (ditto below)

File: examples/sql_bank/main.go (hunk 6 above)
Comment (vivekmenezes) [inline examples/sql_bank/main.go]: I'd rather align this completely with the KV bank app soon and do that all in the same change.

File: examples/sql_bank/main.go (hunk 7 above)
Comment (vivekmenezes) [inline examples/sql_bank/main.go]: I'd like to do that but it makes the code all ugly. You have to declare all the variables. Its pretty harmless to leave this as is.

File: sql/txn.go (hunk 4 above)
Comment (vivekmenezes) [inline sql/txn.go]: Hmmm, yeah I don't believe it's strictly necessary. I'm taking it out from this PR.

Comment (tschottdorf) [inline sql/txn.go]: Yes, that's correct, assuming that's what we wound up implementing.

File: examples/sql_bank/main.go (hunk 2 above)
Comment (vivekmenezes) [inline examples/sql_bank/main.go]: Great. BTW your suggestion didn't work. I've added a test to demonstrate it. You can modify the output of those tests once you've added the point support.

**Comment by vivekmenezes** [comment]
@tamird let me know if you want me to add anything else to this PR before merging. Thanks

**Comment by tamird** [comment]
the sql_bank binary is still checked in

**Comment by tamird** [comment]
I'd like to see the flag thing addressed and for us not to run a server when talking to postgres, but if you want to do that in the next PR, go ahead.

**Pull request merged**